doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,205 @@
+Incidentally, there's a simpler way to make `foldRight` lazy using Scala's built-in syntax for lazy arguments, but it is not stack safe.
+Interestingly, the `Foldable.forall` and `Foldable.exists` functions in cats are implemented by [converting the foldable structure to list first](https://github.com/typelevel/cats/blob/896f00cd689af20676556d1b57791d7ade5cfa04/core/src/main/scala/cats/Foldable.scala#L677-L686), rather than using `foldRight`.
+
+We can potentially introduce something similar to `Eval` in the Plutus Tx standard library.
+The advantage of `Eval` is that it can be provided entirely in a library, without special compiler support. However, this is also what leads to its disadvantage: `Eval` is more heavyweight to use, since it changes the types of many functions (such as `foldr`), while `nonstrict` has the same type as the identity function, which is possible because it is a magical function that the compiler treats specially.
+
+The takeaway is this: the standard libraries usually do the simple things, e.g., providing a monomorphic, non-short-circuiting version of `foldr` for lists, as well as functions like `all`, `any`, also monomorphically for lists.
+The polymorphic versions of those functions, if exist, are usually provided by third-party libraries like cats.
+
+For Plutus Tx, even with `nonstrict`, it is still useful to provide monomorphic versions of certain functions in the standard library, especially since they can be noticeably cheaper than the polymorphic versions.
+This will be discussed in the next section.
+
+### Monomorphizing Plutus Tx Standard Library
+
+Strict languages do not lend themselves to code reuse, as there is a high cost to use abstractions such as higher-order functions and parametric polymorphism.
+Polymorhpic functions taking `Foldable` or similar constraints, implemented in terms of higher-order functions like `foldr`, are rare in strict languages.
+
+For Plutus Tx, there's another reason why monomorphization of the standard library is beneficial: multiple method typeclasses are expensive.
+
+Single method typeclasses are cheap because they have simpler representations in GHC Core - a newtype that can be compiled away.
+On the other hand, multiple method typeclasses exist in GHC Core as actual dictionaries and need to be compiled into product types in PIR.
+
+This makes monomorphization the only choice in many cases. Consider `foldr` and `foldMap`.
+They can be implemented in terms of each other, so technically, only one of them needs to be in the `Foldable` class.
+But certain types have more efficient direct implementations of `foldr`; the same for `foldMap`.
+No matter which of them we move out of `Foldable`, it won't work optimally for certain types.
+Keeping both in `Foldable` will cause `Foldable` to be a multiple method typeclass, which is expensive in and of itself, as explained above.
```

Comment:
I'm going to be annoying and say: we should actually do budget benchmarks of the two versions and see. It's not obvious to me how much difference it makes!

Reply:
We already have proof that multiple method classes are expensive than single-method ones, unless I misunderstood what you meant by "the two versions"?

Reply:
I don't doubt they're more expensive, but how much?

Reply:
I fee like the discussion on the cost of multiple-method typeclasses is independent of the evaluation strategy discussion, so I removed it in the new draft.